### PR TITLE
Floating filter labels

### LINKED ISSUE #4328: Multiple rows dashboard filters
We have highly dimenensional data and often add several filters in our dashboards. Unfortunaltly after a certain number of filters it starts overflowing on the right. It would be better if it could extend over multiple rows if needed. 
Another possibility would be an option to have a vertical filter bar instead of horizontal. (looks better with 5-10+ filters)

### PR DESCRIPTION
**Pull request by salsakran**
Fixed #4328

### CODE REVIEW

**Comment by attekei** [comment]
Will fix #2963 as well. Created based on this design by @mazameli:

screen shot 2017-03-18 at 11 05 03

My thoughts:

- I haven't added field type icons next to field names yet, but currently I'm in favour of adding them. They reduce ambiguity about the field type and thus make it more clear what happens after you click a field.

- Full screen has had its own UI elements for showing active filters, but in this PR it uses the same filter field UI element as other parts of dashboard. It feels pretty big and bold in the full-screen mode, maybe we should keep using the same element but apply some extra styling when in full-screen mode?

- For simple text fields (when editing value of an ID filter, when editing the name of a filter) I'm not totally confident is it clear enough that the only way to save a value is to press Return. Losing the focus of a text field discards its value.

- Is implementing [an animation like this](https://codepen.io/simoberny/pen/gpddYQ) worth the effort?

**Comment by mazameli** [comment]
> For simple text fields (when editing value of an ID filter, when editing the name of a filter) I'm not totally confident is it clear enough that the only way to save a value is to press Return. Losing the focus of a text field discards its value.

I agree that we can improve this. Atte and I spoke and thought we could implement a little hint icon like this:

screen shot 2017-03-20 at 9 39 22 am

Ideally the icon itself could be a button that would commit the input, but that's an enhancement.

**Comment by tlrobinson** [comment]
BTW in #4335 it was suggested that parameters should commit on blur as well. Did we come to a conclusion on that?

**Comment by mazameli** [comment]
> BTW in #4335 it was suggested that parameters should commit on blur as well. Did we come to a conclusion on that?

A few considerations: 
1. we're adding in a little enter icon to help communicate how to commit the input, which should hopefully make the need to hit return clearer
2. Commit-on-blur is potentially dangerous if your dashboard cards take a long time to load. (The latter isn't a consideration for SQL params; they don't auto-reload on commit.) 
3. The best argument I've heard is that commit-on-blur would make tabbing between several filter input boxes feasible, but this is possibly edge-casey? (i.e. how often are users going to have several input box filters in a row?)
4. I don't think commit-on-blur feels quite right from an interaction perspective. That's not how search boxes anywhere else in the world behave, and these input filters are sort of search boxes in that they don't live-update.

So, I thought I was in favor of commit-on-blur, but I think I'm against it again/still.

**Comment by attekei** [comment]
I added the icons indicating filter type (time/location/id/category) as in the original design. Next I will finish the animations and make sure that the filter widgets look sane also in the template tag editor of SQL query builder.

I will leave the commit-on-blur logic as it is.

File: frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx [hunk 1]
```diff
@@ -54,52 +57,88 @@ export default class ParameterValueWidget extends Component {
...
     }
 
+    static getParameterIconName(parameterType) {
+        if (parameterType.search(/date/) !== -1) return "calendar";
+        if (parameterType.search(/location/) !== -1) return "location";
+        if (parameterType.search(/id/) !== -1) return "label";
+        return "clipboard";
+    }
+
+    constructor(props) {
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx]: FYI I just discovered you can replace the whole constructor with `state = { isFocused: false }` if all you're doing is setting initial state.

File: frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx [hunk 2]
```diff
@@ -54,52 +57,88 @@ export default class ParameterValueWidget extends Component {
...
+
     render() {
-        const { parameter, value, values, setValue, isEditing, placeholder, noReset, commitImmediately } = this.props;
+        const {parameter, value, values, setValue, isEditing, placeholder, isFullscreen,
+               noReset, commitImmediately, focusChanged: parentFocusChanged} = this.props;
 
         let hasValue = value != null;
 
         let Widget = ParameterValueWidget.getWidget(parameter, values);
+        const self = this;
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx]: `self = this` tends to be unnecessary with in ES6 with arrow functions.

File: frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx [hunk 3]
```diff
@@ -54,52 +57,88 @@ export default class ParameterValueWidget extends Component {
...
 
-        if (Widget.noPopover) {
-            return (
-                <div className={cx(S.parameter, S.noPopover, { [S.selected]: hasValue })}>
-                    <Widget value={value} values={values} setValue={setValue} isEditing={isEditing} commitImmediately={commitImmediately} />
-                    { hasValue && !noReset &&
-                        <Icon name="close" className="flex-align-right cursor-pointer" size={12} onClick={(e) => {
+        function focusChanged(isFocused) {
+            if (parentFocusChanged) parentFocusChanged(isFocused);
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/parameters/ParameterValueWidget.jsx]: Prefer always including brackets and newlines even if it's a little longer.

File: frontend/src/metabase/dashboard/components/Dashboard.jsx [hunk 4]
```diff
@@ -228,7 +228,7 @@ export default class Dashboard extends Component {
         if (dashboard && dashboard.parameters && dashboard.parameters.length) {
             parameters = (
                 <Parameters
-                    className="ml1"
+                    className=""
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/Dashboard.jsx]: Should be able to just remove this.

File: frontend/src/metabase/dashboard/components/parameters/widgets/TextWidget.jsx [hunk 5]
```diff
@@ -32,7 +35,15 @@ export default class TextWidget extends Component {
...
+        const defaultPlaceholder = this.state.isFocused ? "" : (this.props.placeholder || "Enter a value...");
+        const self = this;
+
+        function focusChanged(isFocused) {
+            if (parentFocusChanged) parentFocusChanged(isFocused);
+            self.setState({isFocused})
+        }
+
         return (
             <input
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/components/parameters/widgets/TextWidget.jsx]: I think we could make this `<input>` have a transparent background so it looks better in full screen etc

File: frontend/src/metabase/dashboard/containers/ParameterWidget.jsx [hunk 6]
```diff
@@ -49,96 +54,120 @@ export default class ParameterWidget extends Component {
...
         const isEditingParameter = editingParameter && editingParameter.id === parameter.id;
 
-        let containerClassName = S.container;
-        let nameClassName = S.name;
-        if (isQB) {
-            containerClassName = "pl2 GuiBuilder-section flex align-center";
-            nameClassName = "GuiBuilder-section-label Query-label";
-        }
+        const containerClassName = cx(S.container);
+        const self = this;
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/containers/ParameterWidget.jsx]: No need for `self = this`

File: frontend/src/metabase/dashboard/containers/ParameterWidget.jsx [hunk 7]
```diff
@@ -49,96 +54,120 @@ export default class ParameterWidget extends Component {
...
         const isEditingDashboard = isEditing;
         const isEditingParameter = editingParameter && editingParameter.id === parameter.id;
 
-        let containerClassName = S.container;
-        let nameClassName = S.name;
-        if (isQB) {
-            containerClassName = "pl2 GuiBuilder-section flex align-center";
-            nameClassName = "GuiBuilder-section-label Query-label";
-        }
+        const containerClassName = cx(S.container);
```
Comment (tlrobinson) [inline frontend/src/metabase/dashboard/containers/ParameterWidget.jsx]: This is sort of redundant. I'd just inline `S.container` below.

**Review by tlrobinson — CHANGES_REQUESTED** [review]
Looking good. The only issues I've noticed (besides code style) are

1. input has a white background. I think we can make the TextWidget <input> background transparent

2. The default picker in SQL variables pane is missing a border and background. Try matching the other elements on the pane (`AdminSelect p1 text-bold text-grey-4 bordered border-med rounded` or similar)

**Comment by attekei** [comment]
Thanks @tlrobinson! I just noticed that clicking the removal button of default picker in SQL variables pane causes an error, will fix that too.

**Review by tlrobinson — APPROVED** [review]

**Pull request merged**